### ISSUE TITLE
The system cannot open the device or file specified.

### ISSUE THREAD

**Issue by Blank-c** [resourcelib/resourcelib#86]
```cs
// IconPath: ./icon.ico (Exists)
// OutputPath: ./test.exe (Exists)
IconFile iconFile = new IconFile(IconPath);
IconDirectoryResource iconDirectoryResource = new IconDirectoryResource(iconFile);
iconDirectoryResource.SaveTo(OutputPath);
```

I am using this code to add icon to a simple program but this doesn't works 9/10 times. I had turned off antivirus but it still doesn't work. The application is ran with administrator permissions too but its still the same result.

**Comment by icnocop** [comment]
What is the full exception including the stack trace?

**Comment by icnocop** [comment]
You can also try to use full paths instead of relative paths to see if that makes any difference.

**Comment by icnocop** [comment]
You can also try batch updating resources if you're updating resource in a loop.
See https://github.com/resourcelib/resourcelib/issues/37 where it also mentions the error "The system cannot open the device or file specified"

**Comment by Blank-c** [comment]
> What is the full exception including the stack trace?

Yes I had tried that too. Still the same error.


> What is the full exception including the stack trace?

I'll check it when I get time asap.



> You can also try batch updating resources if you're updating resource in a loop. See #37 where it also mentions the error "The system cannot open the device or file specified"

I am not really sure, how I can do that. I'd be glad if you can guide me. Thanks.

**Comment by Blank-c** [comment]
I am just reading versioninfo from the file, updating version info, saving back to the file and then using the code to update icon and saving to the same file.

**Comment by icnocop** [comment]
This may be an issue because of the dimensions of the ico.
See https://github.com/resourcelib/resourcelib/issues/77#issuecomment-1272205934

**Comment by Blank-c** [comment]
Like I said 9/10 times it doesn't work but it works sometimes (I am using the same ico)

**Comment by hakito** [comment]
Have you checked if any other process has the file opened?

**Comment by Blank-c** [comment]
Yes I have. I also had turned off windows defender.

**Comment by Blank-c** [comment]
![image](https://user-images.githubusercontent.com/94945186/230720836-65839b28-5467-4c1f-9b8c-42d9652d669e.png)

### Above red line = before turning off AV
### Below red line = after turning off AV.


__Some cases I tested:__
1) Target file, icon, main program in the same directory: Working (✔️).
2) Icon and main program in same directory and target file in another: Error (❌).
3) Target file and main program in same directory and icon in another: Working (✔️).
4) Main program in a directory and target file + icon in another: Error (❌).
5) Every file in different directory: Error (❌).

So basically it points to this: the target file and the main program should be in the same directory for this to work
```cs
using System.IO;
using Vestris.ResourceLib;

namespace test{
    public class Program
    {
        private const string TargetExe = "file_1.exe"; // Modified this according to different cases
        private const string IconPath = "icon.ico"; // Modified this according to different cases

        private const string OriginalExe = "file.exe";

        public static void Main(string[] args)
        {
            if (File.Exists(TargetExe))
            {
                File.Delete(TargetExe);
            }

            File.Copy(OriginalExe, TargetExe);

            IconFile iconFile = new IconFile(IconPath);
            IconDirectoryResource iconDirectoryResource = new IconDirectoryResource(iconFile);
            iconDirectoryResource.SaveTo(TargetExe);
        }
    }
}
```

**Comment by Blank-c** [comment]
In the screenshot, I ran the same program multiple times (without rebuilding).

**Comment by hakito** [comment]
How did you test that no other process is accessing the file?

**Comment by Blank-c** [comment]
I used IObit unlocker

**Comment by hakito** [comment]
I don't know this tool but it sounds like just an unlock tool in case the file is locked. Can you use procmon to monitor file access?

**Comment by Blank-c** [comment]
It also tells you what program is locking the file. Also, I don't think of any reason why some program will lock a file I just copied.

**Comment by dblock** [comment]
Explorer locks the file to read the updated icon so it can display it.

**Comment by dblock** [comment]
@Blank-c No idea. You should confirm which process is causing this, first. I am only guessing.

**Comment by Blank-c** [comment]
> ![image](https://user-images.githubusercontent.com/94945186/230720836-65839b28-5467-4c1f-9b8c-42d9652d669e.png)
> 
> ### Above red line = before turning off AV
> ### Below red line = after turning off AV.
> **Some cases I tested:**
> 
> 1. Target file, icon, main program in the same directory: Working (✔️).
> 2. Icon and main program in same directory and target file in another: Error (❌).
> 3. Target file and main program in same directory and icon in another: Working (✔️).
> 4. Main program in a directory and target file + icon in another: Error (❌).
> 5. Every file in different directory: Error (❌).
> 
> So basically it points to this: the target file and the main program should be in the same directory for this to work
> 
> ```cs
> using System.IO;
> using Vestris.ResourceLib;
> 
> namespace test{
>     public class Program
>     {
>         private const string TargetExe = "file_1.exe"; // Modified this according to different cases
>         private const string IconPath = "icon.ico"; // Modified this according to different cases
> 
>         private const string OriginalExe = "file.exe";
> 
>         public static void Main(string[] args)
>         {
>             if (File.Exists(TargetExe))
>             {
>                 File.Delete(TargetExe);
>             }
> 
>             File.Copy(OriginalExe, TargetExe);
> 
>             IconFile iconFile = new IconFile(IconPath);
>             IconDirectoryResource iconDirectoryResource = new IconDirectoryResource(iconFile);
>             iconDirectoryResource.SaveTo(TargetExe);
>         }
>     }
> }
> ```

Would you mind checking if these cases apply for your systems as well?

**Comment by icnocop** [comment]
Make sure the output directory, where you want to save to, already exists.

**Comment by Blank-c** [comment]
> Make sure the output directory, where you want to save to, already exists.

ye it already does.

**Comment by icnocop** [comment]
Can you try to run the unit tests
https://github.com/resourcelib/resourcelib/blob/0f516aae5f548a56c218e403e2c8bf98abd79162/Source/ResourceLibUnitTests/IconResourceTests.cs#L37
and
https://github.com/resourcelib/resourcelib/blob/0f516aae5f548a56c218e403e2c8bf98abd79162/Source/ResourceLibUnitTests/IconResourceTests.cs#L71 
with your files instead of those in the unit test to see the output of the test?

**Comment by ramiabughazaleh** [comment]
I have noticed this happen when I have Windows Explorer open in the same directory as the output file.
Once I close Windows Explorer, the error no longer occurs.

**Comment by ramiabughazaleh** [comment]
Here are some references:
https://stackoverflow.com/q/32727996/90287
https://stackoverflow.com/a/17697004/90287

**Comment by ramiabughazaleh** [comment]
It seems the file handle returned by `BeginUpdateResource` is opened allowing other processes to read it.

I'm not sure if ResourceLib can avoid using `BeginUpdateResource`, `UpdateResource`, and `EndUpdateResource` and instead use something else which can open the file and deny read access to other processes, to prevent other applications like Windows Explorer and anti-virus programs from trying to read the file until ResourceLib has completely finished updating the resources in the file.

For example, the .NET repository now has a class with managed code to update resources, but I'm not sure if we can use or adapt it to our needs and completely replace the Win32 APIs:
https://github.com/dotnet/runtime/blob/main/src/installer/managed/Microsoft.NET.HostModel/ResourceUpdater.cs

**Comment by hakito** [comment]
I don't think that switching to something else solves this issue. The problem is that another process (explorer) is blocking the access and the resourcelib has no access to the file. 

I think the probability for the issue could increase when multiple icons are tried to be saved at once. Ther is a [loop](https://github.com/resourcelib/resourcelib/blob/0f516aae5f548a56c218e403e2c8bf98abd79162/Source/ResourceLib/DirectoryResource.cs#L100) wich will call `BeginUpdateResource` for each icon.

I think this issue could probably be addressed by (temporarily) changing the file extension to something, that explorer will ignore. After the write operations rename it back.

**Comment by dblock** [comment]
> I think this issue could probably be addressed by (temporarily) changing the file extension to something, that explorer will ignore. After the write operations rename it back.

That would be a good test to try out. Obviously I don't think renaming a file before editing it is something ResourceLib should do ;)

**Comment by hakito** [comment]
You are probably right. It's something that the user can do by himself. Maybe it should just go to the documentation as  a workaround? 

@Blank-c Could you try if this would bypass the problem on your system?
